### DISCUSSION TITLE
Rotation_at(): Produce rotation matrix from ICRS to topocentric frame at Geographic Position

### DISCUSSION THREAD

**Discussion by concord647** [skyfielders/python-skyfield#920] [Q&A] [answered]
Hello again, 

I need a way to generate the rotation matrix from ICRS to a (topocentric) coordinate system (that may or may not be defined as shown in the drawing) defined at a greographic location (i.e. a Ground Station) for specific times. 

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/b4d3a377-2b65-454c-b7e6-1bc9d06bf2e2)

I am looking at Skyfield's rotation_at()-method, but I don't fully understand it. Say a produce a geographic position like this:

```
GS_Lat = 34.22475000  # North
GS_Lon = 118.05591667  # West
GS_Alt = 1728  # meter

GS = wgs84.latlon(GS_Lat * N, GS_Lon * W, elevation_m=GS_Alt)
```

I can then produce a rotation via rotation_at for a specific time:

```
ts = load.timescale()
timenow = ts.now()

# Geocentric position in ICRS at timenow
position_now = GS.at(timenow)

# Rotation ICRS --> what frame??
rotation_now = GS.rotation_at(timenow)
```

This produces some rotation matrix from ICRS --> GS. What I am slightly confused about is the geographic coordinate system. How is this defined, i.e. what exactly is this rotation matrix rotating the ICRS frame to?

Thanks a lot!

**Comment by brandon-rhodes** [comment, marked as answer]
Here's how the documentation defines that reference frame:

https://rhodesmill.org/skyfield/coordinates.html#horizonal-coordinates

Is there a way the title of that section in the docs could be tweaked or expanded to have made it easier for you to find? Should "altitude" or "azimuth" be added to the text of the title, for example?

**Reply by concord647** [reply]
I did actually see that section in the docs! What confused me a little was the fact that in the section "Rotation Matrices", rotation_at() is applied on an ecliptic_frame-object rather that a position-object, so I wasn't sure if I made correct use of that function - plus, I tried to plot the rotation to the resulting coordinate system on a geodetic position and the result did not match my expectation (set with the Horizonal system) at all. I did this for a location on the equator (black points; one location per hour for a day) and expected one axis to point north and one perpendicular to the local horizon, instead I got this:

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/a9680a52-35b6-48d2-a425-30728c5c4b84)

I am not sure if I made a mistake plotting the coordinate systems here or if my general understanding of the matter is flawed. This is the code is used:

```
from skyfield.api import load, wgs84, N, S, E, W
from skyfield import api, units, framelib
from skyfield.functions import T

import matplotlib.pyplot as plt
import numpy as np
from scipy.spatial.transform import Rotation as R

def plot_rotated_axes(ax, r, name=None, offset=(0, 0, 0), scale=2e6):

    colors = ("#FF6666", "#005533", "#1199EE")  # Colorblind-safe RGB
    labels = ('x', 'y', 'z')
    loc = np.array([offset, offset])

    for i, (axis, c, ax_label) in enumerate(zip((ax.xaxis, ax.yaxis, ax.zaxis), colors, labels)):
        axlabel = ax_label
        axis.set_label_text(axlabel)
        axis.label.set_color(c)
        axis.line.set_color(c)
        axis.set_tick_params(colors=c)
        line = np.zeros((2, 3))
        line[1, i] = scale
        line_rot = r.apply(line)
        line_plot = line_rot + loc
        ax.plot(line_plot[:, 0], line_plot[:, 1], line_plot[:, 2], c)
        text_loc = line[1]*1.2
        text_loc_rot = r.apply(text_loc)
        text_plot = text_loc_rot + loc[0]
        ax.text(*text_plot, axlabel.upper(), color=c,
                va="center", ha="center")
    # ax.text(*offset, name, color="k", va="center", ha="center",
            # bbox={"fc": "w", "alpha": 0.8, "boxstyle": "circle"})

## DEFINE GEODETIC OGS COORDINATE SYSTEM
GS_Lat = 0
GS_Lon = 0
GS_Alt = 0

GS = wgs84.latlon(GS_Lat * N, GS_Lon * W, elevation_m=GS_Alt)

ts = load.timescale()
timenow = ts.now()

# Geocentric position in ICRS at timenow
position_now = GS.at(timenow)

# Rotation ICRS --> what frame??
rotation_now = GS.rotation_at(timenow)

print(position_now.position.m)

## Test: Get ground station position in ICRS over full day and plot in 3D
# define start and end time in julian datum
time_start = ts.tt(year=2023, month=10, day=24)
time_end = ts.tt(year=2023, month=10, day=25)
steps = 24

time_vector = ts.linspace(time_start, time_end, num=steps)

# get position of ground station at time_vector
position_vector_fullday = GS.at(time_vector)

# plot
fig = plt.figure()
ax = fig.add_subplot(projection='3d', proj_type="ortho")

# plot ground station positions
ax.scatter(position_vector_fullday.position.m[0], position_vector_fullday.position.m[1], position_vector_fullday.position.m[2], color="k", s=20)

# add sphere to plot
earth_radius = 6371*1000+GS_Alt

# Make data for sphere
u, v = np.mgrid[0:2*np.pi:50j, 0:np.pi:50j]
x = earth_radius*np.cos(u)*np.sin(v)
y = earth_radius*np.sin(u)*np.sin(v)
z = earth_radius*np.cos(v)

# Plot earth surface
ax.plot_surface(x, y, z, rstride=1, cstride=1, color='c', alpha=0.3, linewidth=0)

# plot reference axis
r_none = R.from_matrix(np.identity(3))
plot_rotated_axes(ax, r_none, name="r1", offset=[0,0,0])
plt.show()

# plot horizonal axis at surface points
plot_fraction = 6
plot_indices = np.linspace(0, steps, int(steps/plot_fraction)+1)

for i in np.nditer(plot_indices[0:len(plot_indices)-1]):    
    # r1 = R.from_matrix(np.identity(3))  # translation of CS without rotation
    # r1 = R.from_euler("ZYX", [90, 0, 0], degrees=True)  # single rotation about Z (90°)
    
    # r1 = R.from_matrix(T(GS.rotation_at(time_vector[int(i)])))  # Transposed rotation matrice --> GS --> ICRS ??
    
    r1 = R.from_matrix(GS.rotation_at(time_vector[int(i)]))  # rotation: ICRS --> GS-Frame (Horizonal)

    # euler angles
    euer_angles = r1.as_euler('ZYX', degrees=True)
    print(f'Euler-Angles (ZYX): {euer_angles}')

    # check orthogonality
    print(f'Det of Rotation: {np.linalg.det(GS.rotation_at(time_vector[int(i)]))}')
    
    plot_rotated_axes(ax, r1, name="r1", offset=position_vector_fullday[int(i)].position.m)
    
plt.show()
```

**Reply by brandon-rhodes** [reply]
I hit a small snag with your script, because it stopped at the first `show()`. Is there something I can click or type in the window to tell it that it can move on from the first `show()` so that it can draw the axes and reach the second one? When I close the window, the script exited without the second `show()` having any effect.

I think your only problem is the direction of rotation. The matrix returned by `rotation_at()` rotates vectors _from_ the ICRS _into_ the other reference frame. But you want to do the reverse, and take a line like (1,0,0) in the other reference frame and move it _into_ the IRCS. So try using the `.T` transpose of the matrix. I think the result will line up with what you want!

**Reply by concord647** [reply]
I think some backends block the script-execution on plt.show() for some reason, see e.g. https://stackoverflow.com/questions/28269157/plotting-in-a-non-blocking-way-with-matplotlib
You can delete all plt.show() except the last one - I've only added them for convenience when debuggin/trying to understand what's happening. 3D-Plots in Matplotlib are increadibly slow though, so I've worked on an alternative solution using Plotly. You should be able to open the attached HTML-File (well, it's in a .zip since you can't attach an HTML for some reason...) in a browser and interact with this plot:

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/55cc47d3-72fd-47ab-a796-fb19a929bca4)

I understand using the transposed matrix, that makes sense to me. What is confusing to me though is that the geodetic/topocentric frame seems to rotate around its x-axis, i.e. y and z always point somewhere else. I would have expected that one of them points always north, the other east (or similar), i.e. that the topocentric reference frame is only rotated around the ICRS z-axis.

[Groundstation_2023-12-19-14_14_45.zip](https://github.com/skyfielders/python-skyfield/files/13716204/Groundstation_2023-12-19-14_14_45.zip)

**Reply by brandon-rhodes** [reply]
It sounds frustrating to have to adapt visualizations depending on the speed of the Notebook and the framerate it can support! I look forward to seeing how your plot looks with the transposed matrix in use — hopefully all the frames will line up correctly on your plot.

I'm not quite sure what you mean by "the geodetic/topocentric frame seems to rotate around its x-axis, i.e. y and z always point somewhere else" — if I want to see that happening in the math, what little script might I run?

**Reply by concord647** [reply]
It is - 3D-Plots are a known weakness of Matplotlib though I believe...

What I mean is easier to see if I plot the frames to all positions:

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/7a4799cc-76e1-4ced-929c-228d20d7d4bb)

y and z rotate from position to position. I would have expected (but I may have a wrong understanding here) a "static" local coordinate system on the earth surface that only rotates with the earth's rotation, i.e. around the ICRS Z-axis or in "Lon" in this image:

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/d06b3f68-6fd5-45f3-8196-1f6c219c48bd)

[Groundstation_2023-12-19-15_04_22.zip](https://github.com/skyfielders/python-skyfield/files/13716645/Groundstation_2023-12-19-15_04_22.zip)

**Reply by brandon-rhodes** [reply]
When I made my diagrams, that was a symptom of my applying the matrix backwards. Once I used the transpose, they all lined up north-east-up. Could you check whether you are using the `.T` of the matrix?

**Reply by concord647** [reply]
Yes, sorry I should have stated this: the new plots above (with the colorful earth surface) are with the transposed rotation matrix of the rotation_at()-output, using your T-function:

```
r1 = R.from_matrix(T(GS.rotation_at(time_vector[int(i)])))  # Transposed rotation matrice --> GS --> ICRS
```

**Reply by brandon-rhodes** [reply]
If I had more time at the moment I would clean this up a bit, but here's how I did the rotation in my edited version of your script. It looks like I used my `mxv()` matrix-times-vector call, and used plain-old `.T` because (whoops) I forgot about my `T()` function. See if the same maneuver works for you — this script gives me what look like upright axes; does it work for you too on your system?

```
from skyfield.api import load, wgs84, N, S, E, W
from skyfield import api, units, framelib
from skyfield.functions import T, mxv

import matplotlib.pyplot as plt
import numpy as np
from scipy.spatial.transform import Rotation as R

def plot_rotated_axes(ax, r, rot, name=None, offset=(0, 0, 0), scale=2e6):
    print('k')
    colors = ("#FF6666", "#005533", "#1199EE")  # Colorblind-safe RGB
    labels = ('x', 'y', 'z')
    loc = np.array([offset, offset])

    for i, (axis, c, ax_label) in enumerate(zip((ax.xaxis, ax.yaxis, ax.zaxis), colors, labels)):
        axlabel = ax_label
        axis.set_label_text(axlabel)
        axis.label.set_color(c)
        axis.line.set_color(c)
        axis.set_tick_params(colors=c)
        if rot is None:
            line = np.zeros((2, 3))
            line[1, i] = scale
            line_rot = r.apply(line)
        else:
            v = np.array((0.0, 0.0, 0.0))
            v[i] = scale
            v = mxv(rot.T, v)
            line_rot = np.zeros((2, 3))
            line_rot[1] = v
        line_plot = line_rot + loc
        ax.plot(line_plot[:, 0], line_plot[:, 1], line_plot[:, 2], c)
        text_loc = line_rot[1]*1.2
        text_loc_rot = r.apply(text_loc)
        text_plot = text_loc_rot + loc[0]
        ax.text(*text_plot, axlabel.upper(), color=c,
                va="center", ha="center")
    # ax.text(*offset, name, color="k", va="center", ha="center",
            # bbox={"fc": "w", "alpha": 0.8, "boxstyle": "circle"})

## DEFINE GEODETIC OGS COORDINATE SYSTEM
GS_Lat = 0
GS_Lon = 0
GS_Alt = 0

GS = wgs84.latlon(GS_Lat * N, GS_Lon * W, elevation_m=GS_Alt)

ts = load.timescale()
timenow = ts.now()

# Geocentric position in ICRS at timenow
position_now = GS.at(timenow)

# Rotation ICRS --> what frame??
rotation_now = GS.rotation_at(timenow)

print(position_now.position.m)

## Test: Get ground station position in ICRS over full day and plot in 3D
# define start and end time in julian datum
time_start = ts.tt(year=2023, month=10, day=24)
time_end = ts.tt(year=2023, month=10, day=25)
steps = 24

time_vector = ts.linspace(time_start, time_end, num=steps)

# get position of ground station at time_vector
position_vector_fullday = GS.at(time_vector)

# plot
fig = plt.figure()
ax = fig.add_subplot(projection='3d', proj_type="ortho")

# plot ground station positions
ax.scatter(position_vector_fullday.position.m[0], position_vector_fullday.position.m[1], position_vector_fullday.position.m[2], color="k", s=20)

# add sphere to plot
earth_radius = 6371*1000+GS_Alt

# Make data for sphere
u, v = np.mgrid[0:2*np.pi:50j, 0:np.pi:50j]
x = earth_radius*np.cos(u)*np.sin(v)
y = earth_radius*np.sin(u)*np.sin(v)
z = earth_radius*np.cos(v)

# Plot earth surface
ax.plot_surface(x, y, z, rstride=1, cstride=1, color='c', alpha=0.3, linewidth=0)

# plot reference axis
r_none = R.from_matrix(np.identity(3))
plot_rotated_axes(ax, r_none, None, name="r1", offset=[0,0,0])
#plt.show()

# plot horizonal axis at surface points
plot_fraction = 6
plot_indices = np.linspace(0, steps, int(steps/plot_fraction)+1)
print(plot_indices)

for i in np.nditer(plot_indices[0:len(plot_indices)-1]):    
    # r1 = R.from_matrix(np.identity(3))  # translation of CS without rotation
    # r1 = R.from_euler("ZYX", [90, 0, 0], degrees=True)  # single rotation about Z (90°)
    
    # r1 = R.from_matrix(T(GS.rotation_at(time_vector[int(i)])))  # Transposed rotation matrice --> GS --> ICRS ??
    
    rot = GS.rotation_at(time_vector[int(i)])
    r1 = R.from_matrix(rot)  # rotation: ICRS --> GS-Frame (Horizonal)

    # euler angles
    euer_angles = r1.as_euler('ZYX', degrees=True)
    print(f'Euler-Angles (ZYX): {euer_angles}')

    # check orthogonality
    print(f'Det of Rotation: {np.linalg.det(GS.rotation_at(time_vector[int(i)]))}')
    
    plot_rotated_axes(ax, r1, rot, name="r1", offset=position_vector_fullday[int(i)].position.m)
    
plt.show()
```

**Reply by concord647** [reply]
Thanks Brandon, this does indeed solve the issue:

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/275da0c0-5ead-4254-a45e-9bd2b76208ca)

It took me a while to understand why this produces a different outcome than using Scipy.spatial.transform.Rotation, but I think I got it now: I hadn't realised that the ouput of GS.rotation_at() has a negative determinant, incorporating the "flip" from right-handed to left-handed. R.from_matrix() according to the scipy-docs checks for det=1 and "If the input is not proper orthogonal, an approximation is created using the method described in [[2]](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.from_matrix.html#r4d90f3316501-2)." This approximation changes the transformation in some way.

If I "unflip" the output of GS.rotation_at() with [[1,0,0],[0,-1,0],[0,0,1]], then use the inverse of that with R.apply() for actual rotating , I get a right-handed topocentric coordinate system ("North-West-Up" I guess):

![grafik](https://github.com/skyfielders/python-skyfield/assets/63842120/b5bb769f-60cd-417f-a5c8-3303a4148109)

I completely overlooked that your Horizonal definition is left-handed! Is there a particular reason for that?

Thanks a lot for your support!

**Reply by brandon-rhodes** [reply]
> I completely overlooked that your Horizonal definition is left-handed! Is there a particular reason for that?

Yes, it's because the associated angles — azimuth and altitude — measure "longitude" clockwise around the unit circle instead of counter-clockwise. I have seen systems that instead of returning an altitude angle (also called elevation) instead return a "nadir angle" or somesuch, that's negative for objects above the horizon. But as so many Skyfield users are amateur astronomers, I wanted to go with the standard convention of positive "latitude" being up towards the zenith.

I'm glad we now have your axes correctly oriented! I had never heard of `Scipy.spatial.transform.Rotation` before, but now I'll know about a possible failure condition if I ever see someone using it again on a problem like this.
